src/platforms/esp32/components/avm_builtins/adc_driver.c

```diff
@@ -0,0 +1,771 @@
+}
+
+//
+// Nif Entry/Exit
+//
+
+static void nif_adc_chan_resource_dtor(ErlNifEnv *caller_env, void *obj)
+{
+    UNUSED(caller_env);
+
+    struct ChannelResource *chan_rsrc = (struct ChannelResource *) obj;
+
+    if ((!IS_NULL_PTR(chan_rsrc->cali_handle)) && (chan_rsrc->calibration > ESTIMATED)) {
+#if ADC_CALI_SCHEME_CURVE_FITTING_SUPPORTED
+        ESP_LOGD(TAG, "deregister curve fitting calibration scheme");
+        esp_err_t err = adc_cali_delete_scheme_curve_fitting(chan_rsrc->cali_handle);
+        if (err != ESP_OK) {
+            ESP_LOGE(TAG, "failed to release curve fitting calibration profile");
+        }
+
+#elif ADC_CALI_SCHEME_LINE_FITTING_SUPPORTED
+        ESP_LOGD(TAG, "deregister line fitting calibration scheme");
+        esp_err_t err = adc_cali_delete_scheme_line_fitting(chan_rsrc->cali_handle);
+        if (err != ESP_OK) {
+            ESP_LOGE(TAG, "failed to release line fitting calibration profile");
+        }
+#endif
+    }
+}
+
+static void nif_adc_unit_resource_dtor(ErlNifEnv *caller_env, void *obj)
+{
+    UNUSED(caller_env);
+
+    struct UnitResource *unit_rsrc = (struct UnitResource *) obj;
+
+    if (!IS_NULL_PTR(unit_rsrc->unit_handle)) {
+        esp_err_t err = adc_oneshot_del_unit(unit_rsrc->unit_handle);
+        if (err != ESP_OK) {
+            ESP_LOGE(TAG, "failed to release adc");
+        }
+    }
+#ifdef CONFIG_AVM_ADC2_ENABLE
+    if (!IS_NULL_PTR(unit_rsrc->unit2_handle)) {
+        esp_err_t err = adc_oneshot_del_unit(unit_rsrc->unit2_handle);
+        if (err != ESP_OK) {
+            ESP_LOGE(TAG, "failed to release adc");
+        }
+    }
+#endif
+}
+
+static const ErlNifResourceTypeInit ChannelResourceTypeInit = {
+    .members = 1,
+    .dtor = nif_adc_chan_resource_dtor,
+};
+
+static const ErlNifResourceTypeInit UnitResourceTypeInit = {
+    .members = 1,
+    .dtor = nif_adc_unit_resource_dtor,
+};
+
+static const struct Nif adc_init_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_adc_init
+};
+static const struct Nif adc_deinit_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_adc_deinit
+};
+static const struct Nif adc_acquire_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_adc_acquire
+};
+static const struct Nif adc_release_channel_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_adc_release_channel
+};
+static const struct Nif adc_sample_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_adc_sample
+};
+
+void atomvm_adc_init(GlobalContext *global)
+{
+    ErlNifEnv env;
+    erl_nif_env_partial_init_from_globalcontext(&env, global);
+    adc_channel_resource = enif_init_resource_type(&env, "adc_channel_resource", &ChannelResourceTypeInit, ERL_NIF_RT_CREATE, NULL);
+    adc_unit_resource = enif_init_resource_type(&env, "adc_unit_resource", &UnitResourceTypeInit, ERL_NIF_RT_CREATE, NULL);
+}
+
+const struct Nif *atomvm_adc_get_nif(const char *nifname)
+{
+    TRACE("Locating nif %s ...", nifname);
```

Comment:
I don't know how much this API can be made generic also for non EPS32 devices, if this operation is not straightforward I suggest calling this module esp_adc or something like that, so that would make the transition path to a generic and cross-platform adc API smoother.

Reply:
Very good point. When I first started I thought that stm32 was similar enough that they could use a common API. After closer examination it is quite different, and Pico will be more similar to stm32, but there are enough differences that starting with platform specific interfaces and later writing a common API is probably the best approach here.

Reply:
The module, and Elixir exports have been renamed.